File: Recursion/4.4-Binary-Search-Using-Recursion/c/code.c
#include <stdio.h>

int search(int arr[], int element, int length)
{
    int low = 0;
    int high = length - 1;

    while (low <= high)
    {
        int mid = (low + high) / 2;

        int guess = arr[mid];

        if (guess == element)
        {
            return mid;
        }
        if (element < guess)
        {
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }

    return -1;
}

int main()
{
    int arr[10];

    for (int i = 0; i < 10; i++)
    {
        arr[i] = i + 1;
    }

    int result = search(arr, 5, 10);

    printf("Element found at: %d\n", result);

    return 0;
}

File: Recursion/4.4-Binary-Search-Using-Recursion/python/code.py

def search(arr, el):
    try:
        if not len(arr):
            return None
        mid = (len(arr)-1)//2
        guess = arr[mid]
        if el == guess:
            return mid
        if el < guess:
            return search(arr[:mid], el) 
        else:
            return mid + 1 + search(arr[mid+1:], el)
    except :
        return None

arr = list(range(1,11))
for i in range(0,11):
    print(search(arr, i))
